### ISSUE TITLE
Binding identifier should care about await/yield context

### ISSUE THREAD

**Issue by Kingwl** [microsoft/TypeScript#44459]
# Bug Report

### 🔎 Search Terms
parsing, await, yield

### 🕗 Version & Regression Information

- This changed between versions 3.9.7 and now

### ⏯ Playground Link

[Playground](https://www.typescriptlang.org/play?ts=4.0.5#code/IYZwngdgxgBAZgV2gFwJYHsL3emAKAShgG8AoGGAN2ACcZgB3YVZGAXhgEYBucmKTCHQAbAKYA6YegDmeRs2QFeAX1KlEKDFgBU2dJ0Ik+1OmFSjhAE3ZdeFARCFjJMvGYuWCpVaVCRYGlBomDAARrT4RGQUYqzyLDYATHb8giISUrLxiipqgcE6YbQGUXyxMO5WSSkOThmulZ5qykA)

### 💻 Code

```ts
async function foo () {
  var await = 1; // should be an error, but not
}

async function bar () {
  let await = 2; // error
}

function * foo1() {
  var yield = 1; // should be an error, but not
}

function * bar1() {
  let yield = 2; // error
}
```

### 🙁 Actual behavior

BindingIdentifier in await/yield with `var` keyword does not throw errors.

### 🙂 Expected behavior
It should.

Related: 
- https://github.com/microsoft/TypeScript/pull/43370#issuecomment-855389589
- https://github.com/microsoft/TypeScript/pull/39216 (/cc: @rbuckton )

**Comment by sandersn** [comment]
The problem is that `let await` errors are issued in the parser and `var await` errors are issued in the binder. Syntax errors block reporting of semantic errors.

The parse errors for `let await` are *horrible*, so the right fix is to allow it in the parser and issue the same grammar error as for `var await` in the binder.

Note that this will make the error disappear in plain JS files for now. I hope to fix that soon.

**Comment by Kingwl** [comment]
> Note that this will make the error disappear in plain JS files for now. I hope to fix that soon.

 Sorry, I'm not 100% sure what your means. Does `plain JS files` means js files we emitted?

**Comment by sandersn** [comment]
Sorry, it's a new term. 'plain JS file' means a JS file that is part of a compilation, but is not reporting semantic errors. That is, checkJS is not turned on OR explicitly off, and the file doesn't have `// @ts-check` or `// @ts-no-check` in it. Plain JS files report syntax errors but not semantic errors, so moving a parser error to a grammar error makes it disappear. However, I'm planning to turn on grammar errors for plain JS files in the next release.
